c/dao.c

```diff
@@ -0,0 +1,290 @@
+                                        uint64_t original_capacity,
+                                        uint64_t *calculated_capacity) {
+  int ret;
+  volatile uint64_t len = 0;
+  size_t withdraw_index = 0;
+
+  ret = extract_withdraw_header_index(input_index, &withdraw_index);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  unsigned char deposit_header_buffer[HEADER_SIZE];
+  len = HEADER_SIZE;
+  ret = ckb_load_header(deposit_header_buffer, &len, 0, input_index,
+                        CKB_SOURCE_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  unsigned char withdraw_header_buffer[HEADER_SIZE];
+  len = HEADER_SIZE;
+  ret = ckb_load_header(withdraw_header_buffer, &len, 0, withdraw_index,
+                        CKB_SOURCE_DEP);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  ns(Header_table_t) deposit_header = ns(Header_as_root(deposit_header_buffer));
+  if (deposit_header == NULL) {
+    return ERROR_ENCODING;
+  }
+  ns(Header_table_t) withdraw_header =
+      ns(Header_as_root(withdraw_header_buffer));
+  if (withdraw_header == NULL) {
+    return ERROR_ENCODING;
+  }
+
+  uint64_t deposit_number = ns(Header_number(deposit_header));
+  uint64_t withdraw_number = ns(Header_number(withdraw_header));
+
+  if (withdraw_number <= deposit_number) {
+    return ERROR_INVALID_WITHDRAW_BLOCK;
+  }
+
+  uint64_t windowleft = LOCK_PERIOD_BLOCKS -
+                        (withdraw_number - deposit_number) % LOCK_PERIOD_BLOCKS;
+  uint64_t minimal_since =
+      withdraw_number + MAX(MATURITY_BLOCKS, windowleft) + 1;
+
+  volatile uint64_t input_since = 0;
+  len = 8;
+  ret = ckb_load_input_by_field(((unsigned char *)&input_since), &len, 0, input_index,
+                                CKB_SOURCE_INPUT, CKB_INPUT_FIELD_SINCE);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  if (input_since < minimal_since) {
+    return ERROR_INVALID_WITHDRAW_BLOCK;
+  }
+
+  unsigned char deposit_dao[DAO_SIZE];
+  unsigned char withdraw_dao[DAO_SIZE];
+
+  len = DAO_SIZE;
+  ret = extract_bytes(ns(Header_dao(deposit_header)), deposit_dao, &len);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (len < 8) {
+    return ERROR_ENCODING;
+  }
+  len = DAO_SIZE;
+  ret = extract_bytes(ns(Header_dao(withdraw_header)), withdraw_dao, &len);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (len < 8) {
+    return ERROR_ENCODING;
+  }
+
+  uint64_t deposit_accumulate_rate = *((uint64_t *)(&deposit_dao[8]));
+  uint64_t withdraw_accumulate_rate = *((uint64_t *)(&withdraw_dao[8]));
+
+  volatile uint64_t occupied_capacity = 0;
+  len = 8;
+  ret = ckb_load_cell_by_field(((unsigned char *)&occupied_capacity), &len, 0,
+                               input_index, CKB_SOURCE_INPUT,
+                               CKB_CELL_FIELD_OCCUPIED_CAPACITY);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+
+  uint64_t counted_capacity = 0;
+  if (__builtin_usubl_overflow(original_capacity, occupied_capacity,
```

Comment:
ℹ️**Review Information**

Occupied capacity does not count when calculating interest.